libplatsupport/src/plat/quartz64/serial.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <platsupport/serial.h>
+#include "../../chardev.h"
+
+#define RHR         0x00
+#define THR         0x00
+#define IER         0x04
+#define LSR         0x14
+#define RHR_MASK    MASK(8)
+#define IER_RHRIT   BIT(0)
+#define LSR_TXFIFOE BIT(5)
+#define LSR_RXFIFOE BIT(0)
+
+#define REG_PTR(base, off)     ((volatile uint32_t *)((base) + (off)))
+
+int uart_getchar(ps_chardevice_t *d)
+{
+    int ch = EOF;
+
+    if (*REG_PTR(d->vaddr, LSR) & LSR_RXFIFOE) {
+        ch = *REG_PTR(d->vaddr, RHR) & RHR_MASK;
+    }
+    return ch;
+}
+
+int uart_putchar(ps_chardevice_t* d, int c)
+{
+    while (!(*REG_PTR(d->vaddr, LSR) & LSR_TXFIFOE)) {
+        continue;
+    }
+    *REG_PTR(d->vaddr, THR) = c;
+    if (c == '\n' && (d->flags & SERIAL_AUTO_CR)) {
+        uart_putchar(d, '\r');
```

Comment:
This will print `\n\r` but don't we want to print `\r\n` (CR LF) actually? This bug exists on other platform also and keep spreading with the copy/past of code. This should ne

```
if (c == '\n' && (d->flags & SERIAL_AUTO_CR)) {
        uart_putchar(d, '\r');
}
while (!(*REG_PTR(d->vaddr, LSR) & LSR_TXFIFOE)) {
    /* busy waiting */
}
*REG_PTR(d->vaddr, THR) = c;
```

Reply:
Added this change